src/sql/src/query_model/rewrite_engine.rs

```diff
@@ -0,0 +1,969 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::query_model::{
+    BoxId, BoxScalarExpr, BoxType, ColumnReference, DistinctOperation, Model, QuantifierId,
+    QuantifierSet, QuantifierType, Select,
+};
+use itertools::Itertools;
+use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
+
+/// Rule type
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum RuleType {
+    TopBoxOnly,
+    PreOrder,
+    PostOrder,
+}
+
+/// Trait for rewrite rules
+pub trait Rule {
+    fn name(&self) -> &'static str;
+
+    fn rule_type(&self) -> RuleType;
+
+    /// Whether the action should be fired for the given box.
+    /// This method is not allowed to modify the model in any way.
+    fn condition(&mut self, model: &Model, box_id: BoxId) -> bool;
+
+    /// Invoked immediately after `condition` if it returned true.
+    fn action(&mut self, model: &mut Model, box_id: BoxId);
```

Comment:
Should this return `BoxId`s of new boxes created by the action?

We need to update unique keys after modifying a box, but we probably should also update unique keys of newly created boxes. It seems that returning BoxId's created by the action is more efficient than going down the entire graph.